scalyr_agent/builtin_monitors/journald_monitor.py

```diff
@@ -0,0 +1,379 @@
+
+            if skip_to_end:
+                # seek to the end of the log
+                # NOTE: we need to back up a single item, otherwise journald returns
+                # random entries
+                self._journal.seek_tail()
+                self._journal.get_previous()
+
+            # configure polling of the journal file
+            self._poll = select.poll()
+            mask = self._journal.get_events()
+            self._poll.register( self._journal, mask )
+        except Exception, e:
+            global_log.warn( "Failed to reset journal %s\n%s" % (str(e), traceback.format_exc() ))
+
+    def _get_extra_fields( self, entry ):
+        """
+        Build a dict of key->values based on the fields available in the
+        passed in journal entry, and mapped to the keys in the 'journal_fields' config option.
+        """
+        result = {}
+
+        for key, value in self._extra_fields.iteritems():
+            if key in entry:
+                result[value] = str(entry[key])
+
+        if self._id and 'monitor_id' not in result:
+            result['monitor_id'] = self._id
+
+        return result
+
+    def _has_pending_entries(self):
+        """
+        Checks to see if there are any pending entries in the journal log
+        that are ready for processing
+        """
+
+        # do nothing if there is nothing ready to poll
+        if not self._poll.poll( self._poll_timeout ):
+            return False
+
+        # see if there are any entries to process
+        process = journal.NOP
+        try:
+            process = self._journal.process()
+        except Exception, e:
+            # early return if there was an error
+            global_log.warn( "Error processing journal entries: %s" % str(e),
+                             limit_once_per_x_secs=60, limit_key='journald-process-error' )
+
+            return False
+
+        # If the result is a journal.NOP this means that the journal has not changed
+        # since the last call to process(), which means we have no new entries waiting to
+        # be read
+        if process == journal.NOP:
+            return False
+
+        return True
+
+    def _process_entries(self):
+        """
+        Processes all entries in the journal log.
+
+        Note: This function should only be called directly after a call to _has_pending_entries()
+        returns True.
+        """
+        # read all the entries
+        for entry in self._journal:
+            try:
+                msg = entry.get('MESSAGE', '')
+                extra = self._get_extra_fields( entry )
+                self._logger.emit_value('details', msg, extra_fields=extra)
+                self._last_cursor = entry.get('__CURSOR', None)
+            except Exception, e:
+                global_log.warn( "Error getting journal entries: %s" % str(e),
+                                 limit_once_per_x_secs=60, limit_key='journald-entry-error' )
+
+    def gather_sample(self):
```

Comment:
Ok, after reading this in the documentation, it just seems too weird to overload `sample_interval` for polling journald entries.  How about you just create a `journal_poll_interval` config option.  Then, just invoke the monitor's `set_sample_interval` during the configuration?  Then document here that we just rely on `gather_sample` being invoked once ever `journal_poll_interval`.

Sorry, you'll have to update the documentation, but I think this makes more sense for now.